test/sqlsmith/mzcompose.py

```diff
@@ -0,0 +1,175 @@
+    parser.add_argument("--max-joins", nargs="?", default=None, type=int)
+    parser.add_argument("--seed", nargs="?", default=None, type=int)
+    args = parser.parse_args()
+
+    c.up("materialized")
+
+    # Very simple data for our workload
+    c.sql(
+        """
+        CREATE SOURCE tpch
+          FROM LOAD GENERATOR TPCH (SCALE FACTOR 0.00001)
+          FOR ALL TABLES
+          WITH (SIZE = '1');
+
+        CREATE SOURCE auction
+          FROM LOAD GENERATOR AUCTION (SCALE FACTOR 0.01)
+          FOR ALL TABLES
+          WITH (SIZE = '1');
+
+        CREATE SOURCE counter
+          FROM LOAD GENERATOR COUNTER (SCALE FACTOR 0.0001)
+          WITH (SIZE = '1');"""
+    )
+
+    # queries = args.max_queries or 100000
+    runtime = args.runtime or 600
+    max_joins = (
+        args.max_joins or 2
+    )  # https://github.com/MaterializeInc/materialize/issues/2392
+    seed = args.seed or random.randint(0, 2**31 - 1)
+
+    def kill_sqlsmith_with_delay() -> None:
+        time.sleep(runtime)
+        # c.exec("sqlsmith", "bash", "-c", "kill `pidof sqlsmith`")
+        c.kill("sqlsmith", signal="SIGINT")
+
+    killer = Thread(target=kill_sqlsmith_with_delay)
+    killer.start()
+
+    out = c.run(
+        "sqlsmith",
+        # f"--max-queries={queries}",
+        f"--max-joins={max_joins}",
+        f"--seed={seed}",
+        "--log-json",
+        # "--exclude-catalog", # https://github.com/MaterializeInc/materialize/issues/17871
+        f"--target=host=materialized port=6875 dbname=materialize user=materialize",
+        capture=True,
+        check=False,  # We still get back parsable json on failure, so keep going
+    ).stdout
+    data = json.loads(out)
+
+    new_errors: Dict[FrozenSet[Tuple[str, Any]], List[Dict[str, Any]]] = {}
+    for error in data["errors"]:
+        if not is_known_error(error["message"]):
+            frozen_key = frozenset(
+                {x: error[x] for x in ["type", "sqlstate", "message"]}.items()
+            )
+            if frozen_key not in new_errors:
+                new_errors[frozen_key] = []
+            new_errors[frozen_key].append({x: error[x] for x in ["timestamp", "query"]})
+
+    print(f"SQLsmith: {data['version']} seed: {seed} queries: {data['queries']}")
+    for frozen_key, errors in new_errors.items():
+        key = dict(frozen_key)
+        occurences = f" ({len(errors)} occurences)" if len(errors) > 1 else ""
+        print(
+            f"[SQLsmith] {key['type']} {key['sqlstate']}: {key['message']}{occurences}"
+        )
+        if len(errors) > 1:
+            from_time = datetime.fromtimestamp(errors[0]["timestamp"]).strftime(
+                "%H:%M:%S"
+            )
+            to_time = datetime.fromtimestamp(errors[-1]["timestamp"]).strftime(
+                "%H:%M:%S"
+            )
+            print(f"From {from_time} until {to_time}")
+
+        print(f"Query: {errors[0]['query']}")
```

Comment:
I wonder if it will make a difference if you print the shortest from all available queries for a given error. Or they are all equally bad and unreadable?

Reply:
Good idea.